### ISSUE TITLE
ChatGLM官方的masks部分的代码应该是没有问题的？

### ISSUE THREAD

**Issue by xiningnlp** [hiyouga/ChatGLM-Efficient-Tuning#170]
ChatGLM官方的masks部分的代码应该是没有问题的。

下面是hiyouga的代码，看上去应该采取的时left padding。

```python
 def get_attention_masks(self, input_ids: torch.Tensor, device: torch.device) -> torch.Tensor:
        r"""
        Generates attention masks for left-padded sequences.

        Note that ChatGLM assigns False on token to be attended in attention mask. In general settings, it should be True.

        According to: https://huggingface.co/THUDM/chatglm-6b/blob/v1.1.0/modeling_chatglm.py#L680
        """
        batch_size, seq_length = input_ids.size()
        attention_mask = torch.ones((batch_size, seq_length, seq_length), device=device)
        attention_mask.tril_()
        for i, seq in enumerate(input_ids):
            attention_mask[i, :, :(seq == self.tokenizer.bos_token_id).nonzero()[0].item()] = 1 # context
            attention_mask[i, :, :(seq != self.tokenizer.pad_token_id).nonzero()[0].item()] = 0 # padding
        attention_mask.unsqueeze_(1)
        attention_mask = (attention_mask < 0.5).bool()
        return attention_mask

```

清华官方的代码是假定做了right padding的，它只对每个seq的前context个字符做attention。

```python
def get_masks(self, input_ids, device):
        batch_size, seq_length = input_ids.shape
        context_lengths = [seq.tolist().index(self.config.bos_token_id) for seq in input_ids]
        attention_mask = torch.ones((batch_size, seq_length, seq_length), device=device)
        attention_mask.tril_()
        for i, context_length in enumerate(context_lengths):
            attention_mask[i, :, :context_length] = 1
        attention_mask.unsqueeze_(1)
        attention_mask = (attention_mask < 0.5).bool()

        return attention_mask

```

在采取ROPE的情况下，两者其实是等价的吧？因此也无所谓预训练和微调的一致性了吧？

**Comment by yejiahaoye** [comment]
我的理解是在right padding 之后， seq_length 是 padding 之后的长度，所以attention_mask 中包含了padding 的value 是1 并不会0， 所以导致在算attention score 应该是有偏差的

**Comment by hiyouga** [comment]
这里和 left-padding 或 right-padding 没有关系。我这里只是强调 attention mask 变量中的值，一般来说需要 attend 的部分是 1，否则为 0，但是 ChatGLM 将需要 attend 的部分设置为 False(0)，否则为 True(1)，这和一般情况下截然相反。

**Comment by yejiahaoye** [comment]
那您觉得left-padding 与 rigth-padding 效果一样么， gpt 等自回归的模型好像都是说要left-padding

**Comment by hiyouga** [comment]
我认为效果应该不会有太大差别，尤其是对于目前的 rotary embedding 来说。
